onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+      - `blob_size` = `CeilDiv(block_size * bits, 8)`
+
+    For all bits from 2-8, a row of data is tightly packed and represented by uint8_t.
+    The bit packing specified for [4 bit integer types](https://onnx.ai/onnx/technical/int4.html) is followed.
+    The sub-byte values always start from the LSB of the byte.
+
+    - for 2,4,8 bits, 4x2bit,2x4bit,1x8bit are stored in one `uint8_t`.
+      ```
+        4bit example: with 3x4bit values
+        Value A: 4 bits (Bits 0:3 in byte 0)
+        Value B: 4 bits (Bits 4:7 in byte 0)
+        Value C: 4 bits (Bits 0:3 in byte 1)
+        Packed Bit Layout:
+        Byte 0: [ B B B B A A A A ] -> A[0:3], B[4:7]
+        Byte 1: [ . . . .  C C C C] -> C[0:3]
+      ```
+
+    - for 3,5,6,7 bits, 32x3bit,32x5bit,16x6bit,32x7bit are stored in 12x`uint8_t`, 20x`uint8_t`, 12x`uint8_t`,
+      28x`uint8_t` separately. no bits are wasted.
+      ```
+        3bit example: with 3x3bit values
+        Value A: 3 bits (Bits 0:2 in byte 0)
+        Value B: 3 bits (Bits 3:5 in byte 0)
+        Value C: 3 bits (Bits 6:7 in byte 0 + Bit 0 in byte 1)
+        Packed Bit Layout:
+        Byte 0: [ C C B B B A A A ] -> A[0:2], B[3:5], C[6:7]
+        Byte 1: [ . . . . . . . C ] -> C[0]
+      ```
+  The last `uint_8` byte may have some bits unused.
+
+Input `scales` is stored in same type as original type of B(`float32`, `float16`) with shape like:
+`[N * n_blocks_per_col]`
+
+Input `zero_points` is stored as `uint8_t` or the same type as `A`. It has the same packing method as input `B`.
+  - If `zero_points` type is `uint8_t`, its shape is `[N * CeilDiv(n_blocks_per_col * bits, 8)]`.
+  - If `zero_points` has same type as `A`, it's not packed and has the same shape as `scales`, `[N * n_blocks_per_col]`.
+  - If `zero_points` is not provided then zero_points will be set to `2^(bits - 1)`.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    MatMulNBits,
+    23,
+    OpSchema()
+        .SetDoc(MatMulNBits_ver23_doc)
+        .Attr(
+            "K",
+            "Size of each input feature. If not present, should be inferred from input A.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "N",
+            "Size of each output feature. If not present, should be inferred from input B.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "accuracy_level",
+            "The minimum accuracy level of input A, can be: "
+            "0(unset), 1(float), 2(float16), 3(Bfloat16), 4(int8) (default unset). It is used to control how input A "
+            "is quantized or downcast internally while doing computation, for example: 0 means input A will not be "
+            "quantized or downcast while doing computation, 4 means input A can be quantized with the same block_size"
+            "to int8 internally from type T1.",
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+        .Attr(
+            "bits",
```

Comment:
the attribute "bits" alone is not enough to specify type of quantization. For example, with bitnet ternary quantization, it takes 2 bit only for 3 states of -1, 0 1. It may further compress the ternary weight to 3 weight values per 5 nits. I  suggest to add a optional attribute to allow indicating this types of cases. In terms of which string to use, we can discuss.